File: daemon/daemon.js
var redis = require('redis');
var redisClient = redis.createClient();
var RANDOM_QUEUE_NAME = 'queue';

redisClient.on('error', function(err) {
  console.log('Error:: ' + err);
})

redisClient.zadd(RANDOM_QUEUE_NAME, 0.1, 'a');
redisClient.zadd(RANDOM_QUEUE_NAME, 0.2, 'b');
redisClient.zadd(RANDOM_QUEUE_NAME, 0.3, 'c');
redisClient.zadd(RANDOM_QUEUE_NAME, 0.4, 'd');
redisClient.zadd(RANDOM_QUEUE_NAME, 0.5, 'e');
redisClient.zadd(RANDOM_QUEUE_NAME, 0.5, 'f');


redisClient.zcard(RANDOM_QUEUE_NAME, function(err, response) {
  if (+response > 1) {
    redisClient.zrange([RANDOM_QUEUE_NAME, 0, 1], function(err, response) {
      if (Object.prototype.toString.call(response) !== '[object Array]')
        return;

      if (response.length < 2)
        return;

      console.log(response);

      var id1 = response[0];
      var id2 = response[1];

      redisClient.multi()
        .zrem(RANDOM_QUEUE_NAME, id1)
        .zrem(RANDOM_QUEUE_NAME, id2)
        .exec(function(err, response) {
          redisClient.publish(id1, id2);
          redisClient.publish(id2, id1);
        });
    });
  }
});

File: daemon/daemon.py
import random
import time

import redis

rs = redis.Redis()
RANDOM_QUEUE_NAME = 'queue'


def populate():
    """ Populate the queue with some dummy ids. """
    rs.zadd(RANDOM_QUEUE_NAME, 'a', random.random())
    rs.zadd(RANDOM_QUEUE_NAME, 'b', random.random())
    rs.zadd(RANDOM_QUEUE_NAME, 'c', random.random())
    rs.zadd(RANDOM_QUEUE_NAME, 'd', random.random())
    rs.zadd(RANDOM_QUEUE_NAME, 'e', random.random())
    rs.zadd(RANDOM_QUEUE_NAME, 'f', random.random())


def dequeue():
    qlen = rs.zcard(RANDOM_QUEUE_NAME)
    if qlen < 1:
        return

    ids = rs.zrange(RANDOM_QUEUE_NAME, 0, -1)
    print 'ids', ids

    if not ids or len(ids) < 1:
        return

    i = 0
    while (i + 1 < len(ids)):
        id1 = ids[i]
        id2 = ids[i + 1]
        print id1, id2

        p = rs.pipeline(transaction=True)
        p.zrem(RANDOM_QUEUE_NAME, id1)
        p.zrem(RANDOM_QUEUE_NAME, id2)
        p.execute()
        rs.publish(id1, id2)
        rs.publish(id2, id1)
        i += 2


if __name__ == '__main__':
    cnt = 0
    while True:
        cnt += 1
        time.sleep(0.1)
        dequeue()
